### PR TITLE
Bugfix and simplify JSONP callback call when the callback name is fixed by the remote

### PR DESCRIPTION
**Pull request by sergi**
The bug is one in which a callback in a fixed jsonp scenario would be called with an Observable as a first parameter, instead of the data padded in the JSONP.

### CODE REVIEW

File: src/jsonp.js [hunk 1]
```diff
@@ -41,28 +41,22 @@
...
         if (typeof settings.jsonpCallback === 'string') {
           handler = settings.jsonpCallback;
-          prevFn = root[handler];
         }
 
-        settings.url = settings.url.replace('=' + settings.jsonp, '=' + handler);
-
-        root[handler] = function(data) {
-          if (prevFn && typeof prevFn === 'function') {
-            prevFn(observer, data);
```
Comment (sergi) [inline src/jsonp.js]: This was wrong. Observer should not be the first parameter.

**Comment by paulpdaniels** [comment]
The url is being defined before the handler.  
Also this doesn't fix for the scenario in #39, which seems to be because at [this conditional](https://github.com/sergi/RxJS-DOM/blob/0e3f3a8c67100ca9d9449b676a0cd9c41d11065b/src/jsonp.js#L50) only the first handler will be added, subsequent ones will be ignored (in the previous iteration a new callback was being created to wrap the old one but then only proxying the values on).

I think if you change it like so it should work:
```javascript

var existing = root[handler];
root[handler]  = function(data) {
  existing && existing(data);
  defaultCallback(observer, data);
}
```

**Comment by sergi** [comment]
The url being defined before the handler is not a problem, because the url doesn't matter if `jsonpCallback` is set. But I will change it for code clarity, no problem.

The code you propose would call `existing` *and* `defaultCallback` with the data, but we only want to call `existing` if it is truthy. So this would be better, I think:

```javascript
var existing = root[handler];
root[handler] = function(data) {
  if (existing) {
    return existing(data);
  }
  defaultCallback(observer, data);
}
```

What do you think? Although in my tests this still doesn't solve #39.

**Comment by paulpdaniels** [comment]
Why would it not matter?


```javascript

        settings.url = settings.url.replace('=' + settings.jsonp, '=' + handler);
        if (typeof settings.jsonpCallback === 'string') {
          handler = settings.jsonpCallback;
        }


```

If it is set before as in the above code then the url would be
Url:
```
 http://ip.jsontest.com/?callback=JSONPCallback
```
becomes
```
http://myurl.com/something?callback=rxjscallback0
```
meanwhile the callback you declared in root would be whatever the passed in jsonpCallback name is.

Ah ok, I misunderstood the intent.  I had thought you wanted to propagate the first response ala amb() but I see, you want each invocation to return a separate value, so you could potentially have multiple in-flight requests that resolve at different times.  Is that correct?

The problem that I see is that since A (likely) completes first B will still have the reference to A and thus never get called, there needs to be a recursive unwinding of the callbacks
The following snippet works with n subscriptions in my tests, the idea is that each time you execute the handler recursively, the bottom-most item executes defaultCallback and returns true, notifying its parent that _it_ is now the bottom most for the next call, and so on.

```javascript

var existing = root[handler];
root[handler] = function(data) {
  if (existing) {
    existing(data) && (existing = null);
    return false;
  }
  defaultCallback(observer, data);
  return true;
};

```

**Comment by sergi** [comment]
Hi @paulpdaniels,

Thanks for this thorough follow-up, I really appreciate it.

It wouldn't matter because, since the name of the callback is hardcoded in these kind of jsonp documents, the querystring won't have any effect on changing the name.

Tried your code and it makes sense, and works most of the time, but I always get at least one subscription that is only handled by one observer, instead of by the two of them:

Screenshot

The code I am trying with looks like this:

    var quakes = Rx.Observable
        .interval(5000)
        .flatMap(function() {
          console.log('*')
          return Rx.DOM.jsonpRequest({
            url: URL,
            jsonpCallback: 'test_callback'
          });
        })

      quakes.subscribe(function(row) { console.log('A', row) });
      quakes.subscribe(function(row) { console.log('B', row) });

Do you get similar results?

**Comment by paulpdaniels** [comment]
Gah! Recursion is such a cruel mistress...
Forgot to clean up the top level handler before it returns.

Let's try this:
```javascript
var existing = root[handler];
root[handler] = function (data, recursed) {
  if (existing) {
    existing(data, true) && (existing = null);
    return false;
  }
  defaultCallback(observer, data);
  !recursed && (root[handler] = null);
  return true;
};


```

**Comment by sergi** [comment]
Hi @paulpdaniels, sorry for the delay. This seems to work, I will change the code accordingly, thanks!

**Comment by mattpodwysocki** [comment]
@sergi let me know when it is done. Also need to get the tests passing again

**Comment by sergi** [comment]
@mattpodwysocki @paulpdaniels Fixed. Please check the code. I added a test case and the jsonp test file passes completely, although the global RxJS-DOM tests don't, as you seem to know already :wink:

**Pull request merged**